### DISCUSSION TITLE
Seeking Suggestions on Improving Code Efficiency for Retrieving Results from Milvus Collection.

### DISCUSSION THREAD

**Discussion by ZubeenK** [milvus-io/milvus#25258] [Q&A and General discussion]
I am using the below code for retrieving results from milvus collection:

```
batch_size = 8000
out = []
for i in range(0, len(embeddings), batch_size):
    search_embeddings = embeddings3[i:i+batch_size]
    results = search_vectors(collection_name='test', search_vector=search_embeddings, num_cluster_unit_query=2048, topk=2, metric="L2")
    out.append(results)
```
  
  the code for search vectors is below:
```
  def search_vectors(collection_name: str, search_vector: list, num_cluster_unit_query: int, topk: int, metric: str = "L2",):
      collection = Collection(collection_name)
      collection.load()
      search_params = {"metric_type": metric, "params": {"nprobe": num_cluster_unit_query}}
      results = collection.search(
          data=search_vector,
          anns_field="embedding",  # ANNS = approx nearest neighbour search
          param=search_params,
          limit=topk,
          expr=None,
      )
      return results
```

But since my dataset is large (~307000 images), it takes a long time for retrieval and storing results in a list. Can anyone suggest a better way to do this?

**Comment by yhmo** [comment]
The "nprobe"=2048 is too large. With this value, it nearly to be brute-force search.

**Reply by ZubeenK** [reply]
Can you suggest a way for me to set nprobe? Which can give me accuracy with speed

**Reply by yhmo** [reply]
Smaller nprobe value, search performance is better, accuracy is worse.
Larger nprobe value, search performance is worse, accuracy is better.

For IVF index, you can choose the "nlist" value according to this FAQ: https://milvus.io/docs/performance_faq.md#How-to-set-nlist-and-nprobe-for-IVF-indexes

Milvus builds an independent index for each segment, so we need to estimate how many vectors in each segemnt.
From the "segment.maxSize" and vector dimension we can estimate how many vectors in a segment. 
For example, segment.maxSize=512MB, vector dimension=512, then ideally each segment contains 512MB/(512*4bytes)=262144 vectors.
According to the FAQ, we say the best nlist is calculated by "4 × sqrt(n)", "n" is the average vector count of each segment. Then we can know the nlist is 4 × sqrt(262144)=2048.

Now we want to search, the "nprobe" ought to be less than "nlist". If we want to search 10% of the vectors, we can set "nprobe"=200. If the result accuracy is not good, then we can increase "nprobe" to 400. By this way, we can make trade-off between accuracy and performance.

**Reply by ZubeenK** [reply]
The vector dimension in my use case is 2048, so the nlist should be around 1000, right?
Also what does it mean to search 10% of the vectors? Like searching for the closest 10% or searching from 10% of the vectors present in milvus collection. In my use case, I want top 2 results for every query vector, so what should be the value of nprobe for my use case?

**Reply by yhmo** [reply]
dimension=2048, segment.maxSize=512MB, then each segment contains 512MB/(2048 * 4bytes)=65536 vectors,  nlist=4 * sqrt(65536)=1024.

nlist=1024 means the IVF index split the 65536 vectors into 1024 buckets, each bucket has a centroid vector, average there are 64 vectors in each bucket. When you call search(nprobe=100), the index first calculates the target vector with the 1024 centroid vectors and picks up 100 closer buckets. Secondly, it calculates the target vector with all the vectors in the 100 buckets and finds out the closest topk vectors. the vectors in 100 buckets are about 10% of the total number, so we say it only searches 10% of the vectors.
Set nprobe=100, if the result accuracy is acceptable, then nprobe=100 is ok for you. If not, increase nprobe to 200.

**Reply by ZubeenK** [reply]
Thanks @yhmo for explaining, it definitely helps me understanding how to set nlist and nprobe better now.
